File: dots_boxes_mcts/encoding.py
from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from dots_boxes_mcts.game import all_edge_ids, box_edge_ids, new_game, parse_edge_id, state_snapshot

CHANNEL_NAMES = (
    "drawn_edge",
    "current_player_edge",
    "opponent_edge",
    "current_player_box",
    "opponent_box",
    "legal_move",
    "current_player",
    "score_margin",
)


@dataclass(frozen=True)
class EncodedPosition:
    tensor: np.ndarray
    legal_mask: np.ndarray
    action_ids: tuple[str, ...]


def board_shape(rows: int, cols: int) -> tuple[int, int]:
    return 2 * rows - 1, 2 * cols - 1


def action_ids(rows: int, cols: int) -> tuple[str, ...]:
    return tuple(all_edge_ids(rows, cols))


def action_index(edge_id_value: str, rows: int, cols: int) -> int:
    try:
        return action_ids(rows, cols).index(edge_id_value)
    except ValueError as error:
        raise ValueError(f"Unknown edge id for {rows}x{cols} board: {edge_id_value}") from error


def edge_coordinate(edge_id_value: str) -> tuple[int, int]:
    kind, row, col = parse_edge_id(edge_id_value)
    if kind == "h":
        return 2 * row, 2 * col + 1
    return 2 * row + 1, 2 * col


def box_coordinate(row: int, col: int) -> tuple[int, int]:
    return 2 * row + 1, 2 * col + 1


def encode_snapshot(snapshot: dict) -> EncodedPosition:
    rows = int(snapshot["rows"])
    cols = int(snapshot["cols"])
    player = int(snapshot["currentPlayer"])
    opponent = 1 if player == 0 else 0
    height, width = board_shape(rows, cols)
    tensor = np.zeros((len(CHANNEL_NAMES), height, width), dtype=np.float32)

    edge_owners = {edge: owner for edge, owner in snapshot.get("edgeOwners", [])}
    for edge in snapshot.get("edges", []):
        row, col = edge_coordinate(edge)
        owner = edge_owners.get(edge)
        tensor[0, row, col] = 1.0
        if owner == player:
            tensor[1, row, col] = 1.0
        elif owner == opponent:
            tensor[2, row, col] = 1.0

    for row_index, row in enumerate(snapshot.get("boxes", [])):
        for col_index, owner in enumerate(row):
            if owner is None:
                continue
            row_coord, col_coord = box_coordinate(row_index, col_index)
            if owner == player:
                tensor[3, row_coord, col_coord] = 1.0
            elif owner == opponent:
                tensor[4, row_coord, col_coord] = 1.0

    edge_set = set(snapshot.get("edges", []))
    ids = action_ids(rows, cols)
    legal_mask = np.array([edge not in edge_set for edge in ids], dtype=np.float32)
    for edge, is_legal in zip(ids, legal_mask, strict=True):
        if is_legal:
            row, col = edge_coordinate(edge)
            tensor[5, row, col] = 1.0

    tensor[6, :, :] = 1.0 if player == 1 else 0.0
    scores = snapshot.get("scores", [0, 0])
    total_boxes = max((rows - 1) * (cols - 1), 1)
    tensor[7, :, :] = (float(scores[player]) - float(scores[opponent])) / total_boxes

    return EncodedPosition(tensor=tensor, legal_mask=legal_mask, action_ids=ids)


def legal_mask_from_snapshot(snapshot: dict) -> np.ndarray:
    return encode_snapshot(snapshot).legal_mask


def empty_board_snapshot(rows: int, cols: int) -> dict:
    return state_snapshot(new_game(rows=rows, cols=cols))


def box_edges_as_coordinates(row: int, col: int) -> list[tuple[int, int]]:
    return [edge_coordinate(edge) for edge in box_edge_ids(row, col)]

File: static_bot/public/src/game/encoding.js
import { allEdgeIds, parseEdgeId } from "./rules.js";

export const CHANNEL_NAMES = [
  "drawn_edge",
  "current_player_edge",
  "opponent_edge",
  "current_player_box",
  "opponent_box",
  "legal_move",
  "current_player",
  "score_margin",
];

export function boardShape(rows, cols) {
  return { height: 2 * rows - 1, width: 2 * cols - 1 };
}

export function edgeCoordinate(edge) {
  const { kind, row, col } = parseEdgeId(edge);
  if (kind === "h") {
    return { row: 2 * row, col: 2 * col + 1 };
  }
  return { row: 2 * row + 1, col: 2 * col };
}

export function boxCoordinate(row, col) {
  return { row: 2 * row + 1, col: 2 * col + 1 };
}

export function encodeSnapshot(snapshot) {
  const rows = Number(snapshot.rows);
  const cols = Number(snapshot.cols);
  const player = Number(snapshot.currentPlayer);
  const opponent = player === 0 ? 1 : 0;
  const { height, width } = boardShape(rows, cols);
  const channels = CHANNEL_NAMES.length;
  const tensor = new Float32Array(height * width * channels);
  const edgeOwners = new Map((snapshot.edgeOwners || []).map(([edge, owner]) => [edge, owner]));

  for (const edge of snapshot.edges || []) {
    const coord = edgeCoordinate(edge);
    const owner = edgeOwners.get(edge);
    tensor[indexOf(coord.row, coord.col, 0, width, channels)] = 1;
    if (owner === player) {
      tensor[indexOf(coord.row, coord.col, 1, width, channels)] = 1;
    } else if (owner === opponent) {
      tensor[indexOf(coord.row, coord.col, 2, width, channels)] = 1;
    }
  }

  for (let row = 0; row < (snapshot.boxes || []).length; row += 1) {
    for (let col = 0; col < snapshot.boxes[row].length; col += 1) {
      const owner = snapshot.boxes[row][col];
      if (owner === null || owner === undefined) {
        continue;
      }
      const coord = boxCoordinate(row, col);
      if (owner === player) {
        tensor[indexOf(coord.row, coord.col, 3, width, channels)] = 1;
      } else if (owner === opponent) {
        tensor[indexOf(coord.row, coord.col, 4, width, channels)] = 1;
      }
    }
  }

  const edgeSet = new Set(snapshot.edges || []);
  const ids = allEdgeIds(rows, cols);
  const legalMask = new Float32Array(ids.length);
  for (let action = 0; action < ids.length; action += 1) {
    const edge = ids[action];
    if (!edgeSet.has(edge)) {
      const coord = edgeCoordinate(edge);
      legalMask[action] = 1;
      tensor[indexOf(coord.row, coord.col, 5, width, channels)] = 1;
    }
  }

  if (player === 1) {
    fillChannel(tensor, height, width, channels, 6, 1);
  }

  const scores = snapshot.scores || [0, 0];
  const totalBoxes = Math.max((rows - 1) * (cols - 1), 1);
  fillChannel(tensor, height, width, channels, 7, (scores[player] - scores[opponent]) / totalBoxes);

  return { tensor, legalMask, actionIds: ids, height, width, channels };
}

export function indexOf(row, col, channel, width, channels) {
  return (row * width + col) * channels + channel;
}

function fillChannel(tensor, height, width, channels, channel, value) {
  for (let row = 0; row < height; row += 1) {
    for (let col = 0; col < width; col += 1) {
      tensor[indexOf(row, col, channel, width, channels)] = value;
    }
  }
}
